src/Xamarin.Android.Build.Tasks/Utilities/ManagedResourceParser.cs

```diff
@@ -0,0 +1,182 @@
+﻿using System;
+using System.CodeDom;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using System.Xml.Linq;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.Android.Tasks
+{
+	class ManagedResourceParser 
+	{
+		public static TaskLoggingHelper Log { get; set; }
+
+		CodeTypeDeclaration resources;
+		CodeTypeDeclaration layout, ids, drawable, strings, colors, dimension, raw;
+
+		public CodeTypeDeclaration Parse (string resourceDirectory, IEnumerable<string> additionalResourceDirectories, bool isApp, Dictionary<string, string> resourceMap)
+		{
+			if (!Directory.Exists (resourceDirectory))
+				throw new InvalidOperationException ("Specified resource directory was not found: " + resourceDirectory);
+
+			resources = CreateResourceClass ();
+			layout = CreateClass ("Layout");
+			ids = CreateClass ("Id");
+			drawable = CreateClass ("Drawable");
+			strings = CreateClass ("String");
+			colors = CreateClass ("Color");
+			dimension = CreateClass ("Dimension");
+			raw = CreateClass ("Raw");
+
+			foreach (var dir in Directory.GetDirectories (resourceDirectory, "*", SearchOption.TopDirectoryOnly))
+				foreach (var file in Directory.GetFiles (dir, "*.*", SearchOption.AllDirectories)) {
+					ProcessResourceFile (file);
+				}
+			foreach (var dir in additionalResourceDirectories)
+				foreach (var file in Directory.GetFiles (dir, "*.*", SearchOption.AllDirectories)) {
+					ProcessResourceFile (file);
+				}
+			if (colors.Members.Count > 1)
+				resources.Members.Add (colors);
+			if (dimension.Members.Count > 1)
+				resources.Members.Add (dimension);
+			if (drawable.Members.Count > 1)
+				resources.Members.Add (drawable);
+			if (ids.Members.Count > 1)
+				resources.Members.Add (ids);
+			if (layout.Members.Count > 1)
+				resources.Members.Add (layout);
+			if (raw.Members.Count > 1)
+				resources.Members.Add (raw);
+			if (strings.Members.Count > 1)
+				resources.Members.Add (strings);
+
+			return resources;
+		}
+
+		void ProcessResourceFile (string file)
+		{
+			var fileName = Path.GetFileNameWithoutExtension (file);
+			if (string.IsNullOrEmpty (fileName))
+				return;
+			if (fileName.EndsWith (".9", StringComparison.OrdinalIgnoreCase))
+				fileName = Path.GetFileNameWithoutExtension (fileName);
+			var path = Directory.GetParent (file).Name;
+			var ext = Path.GetExtension (file);
+			if (ext == ".xml" || ext == ".axml" && string.Compare (path, "raw", StringComparison.OrdinalIgnoreCase) != 0) {
```

Comment:
Let's design for the future here: we're going to have *lots* of potential file extensions, which all require different parsers. We'll thus either wind up with a *gigantic* `ProcessResourceFile()` method, or...we're going to have lots of "helper methods."

Lots of methods is better than one gigantic method, so let's just force that. :-)

```csharp
switch (ext) {
    case ".xml":
    case ".axml":
        if (string.Compare (path, "raw", StringComparison.OrdinalIgnoreCase) == 0)
            goto case default;
        ProcessXmlResourceFile (...);
        break;
    default:
        break;
}
```